lib/engine/game/g_1870.rb

```diff
@@ -69,15 +78,121 @@ def init_stock_market
+        end
+
+        @hexes.each do |hex|
+          hex.assignments.dup.each do |company, _|
+            if ASSIGNMENT_TOKENS.keys.any?(company)
+              removals[company][:hex] = hex.name
+              hex.remove_assignment!(company)
+            end
+          end
+        end
+
+        removals.each do |company, removal|
+          hex = removal[:hex]
+          corp = removal[:corporation]
+          @log << "-- Event: #{corp}'s #{company_by_id(company).name} token removed from #{hex} --"
+        end
+      end
+
+      def river_company
+        @river_company ||= company_by_id('MRBC')
+      end
+
+      def port_company
+        @port_company ||= company_by_id('GSC')
+      end
+
+      def mp_corporation
+        @mp_corporation ||= corporation_by_id('MP')
+      end
+
+      def ssw_corporation
+        @ssw_corporation ||= corporation_by_id('SSW')
+      end
+
+      def river_corporations
+        [ssw_corporation, mp_corporation]
+      end
+
+      def purchasable_companies(entity = nil)
+        return super unless @phase.name == '1'
+        return [river_company] if [mp_corporation, ssw_corporation].any?(entity)
+
+        []
+      end
+
       def corporation_opts
         { can_hold_above_max: true }
       end
 
-      def sell_shares_and_change_price(bundle)
+      def assignment_tokens(assignment)
+        return "/icons/#{assignment.logo_filename}" if assignment.is_a?(Engine::Corporation)
+
+        super
+      end
+
+      def revenue_for(route, stops)
+        revenue = super
+
+        cattle = 'SCC'
+        revenue += 10 if route.corporation.assigned?(cattle) && stops.any? { |stop| stop.hex.assigned?(cattle) }
+
+        revenue += 20 if route.corporation.assigned?('GSCᶜ') && stops.any? { |stop| stop.hex.assigned?('GSCᶜ') }
+
+        revenue += (route.corporation.assigned?('GSC') ? 20 : 10) if stops.any? { |stop| stop.hex.assigned?('GSC') }
+
+        revenue
+      end
+
+      def sell_shares_and_change_price(bundle, _allow_president_change: true, _swap: nil)
         @round.sell_queue << bundle
 
         @share_pool.sell_shares(bundle)
       end
+
+      def legal_tile_rotation?(_entity, hex, tile)
+        return true unless river_company.abilities(:blocks_crossing_partition)
+
+        (tile.exits & hex.tile.borders.select { |b| b.type == :water }.map(&:edge)).empty? &&
+          hex.tile.partitions.all? do |partition|
+            tile.paths.all? do |path|
+              (path.exits - partition.inner).empty? || (path.exits - partition.outer).empty?
+            end
+          end
+      end
+
+      def upgrades_to?(from, to, _special = false)
+        return to.name == '170' if from.color == :green && P_HEXES.any?(from.hex.name)
+
+        return to.name == '171K' if from.color == :brown && from.hex.name == 'B11'
+        return to.name == '172L' if from.color == :brown && from.hex.name == 'C18'
+
+        super(from, to, false, to.cities.size.positive? && %i[yellow green].any?(to.color))
```

Comment:
to.cities.any? && [].include?(to.color)

Reply:
I love how I forgot that `any?` existed when it actually was what I wanted to do :man_facepalming: